File: frontend/src/inboxTabOrder.js
/** Orden de chips de bandeja (Contactos · Grupos · No leídos · Favoritos). */

export const INBOX_TAB_IDS = ['contacts', 'groups', 'unread', 'favorites'];

export const INBOX_TAB_LABELS = {
  contacts: 'Contactos',
  groups: 'Grupos',
  unread: 'No leídos',
  favorites: 'Favoritos',
};

const LEGACY_MAP = {
  all: 'contacts',
  contacts: 'contacts',
  groups: 'groups',
  unread: 'unread',
  favorites: 'favorites',
};

export function inboxTabOrderStorageKey(userId) {
  return `tacticalptx_chat_tab_order_${userId || 'anon'}`;
}

export function normalizeInboxTabOrder(raw) {
  const seen = new Set();
  const out = [];
  for (const id of Array.isArray(raw) ? raw : []) {
    const mapped = LEGACY_MAP[String(id)] || null;
    if (!mapped || seen.has(mapped)) continue;
    seen.add(mapped);
    out.push(mapped);
  }
  for (const id of INBOX_TAB_IDS) {
    if (!seen.has(id)) out.push(id);
  }
  return out;
}

export function loadInboxTabOrder(userId) {
  try {
    const raw = JSON.parse(localStorage.getItem(inboxTabOrderStorageKey(userId)) || 'null');
    return normalizeInboxTabOrder(raw);
  } catch {
    return [...INBOX_TAB_IDS];
  }
}

export function saveInboxTabOrder(userId, order) {
  try {
    localStorage.setItem(
      inboxTabOrderStorageKey(userId),
      JSON.stringify(normalizeInboxTabOrder(order))
    );
  } catch {
    /* ignore */
  }
}

export function moveInboxTab(order, fromId, toId) {
  const next = normalizeInboxTabOrder(order);
  const from = next.indexOf(fromId);
  const to = next.indexOf(toId);
  if (from < 0 || to < 0 || from === to) return next;
  next.splice(from, 1);
  next.splice(to, 0, fromId);
  return next;
}

/** 0 en línea → 1 ausente → 2 fuera de línea → 3 desconectado (paridad APK). */
function presenceSortRank(presence, online) {
  const p = String(presence || '')
    .toLowerCase()
    .trim();
  if (p === 'online' || p === 'radio' || p === 'active' || p === 'foreground') return 0;
  if (p === 'away' || p === 'background' || p === 'service') return 1;
  if (p === 'stale') return 2;
  if (p === 'offline') return 3;
  return online ? 0 : 3;
}

/** Comparador Contactos: presencia → grado militar → último mensaje. */
export function compareContactRows(a, b) {
  const ar = presenceSortRank(a.presence, a.online);
  const br = presenceSortRank(b.presence, b.online);
  if (ar !== br) return ar - br;
  const ag = Number.isFinite(a.gradeSortOrder) ? a.gradeSortOrder : 999999;
  const bg = Number.isFinite(b.gradeSortOrder) ? b.gradeSortOrder : 999999;
  if (ag !== bg) return ag - bg;
  const ta = a.at ? new Date(a.at).getTime() : 0;
  const tb = b.at ? new Date(b.at).getTime() : 0;
  return tb - ta;
}

export function compareByLastMessage(a, b) {
  const ta = a.at ? new Date(a.at).getTime() : 0;
  const tb = b.at ? new Date(b.at).getTime() : 0;
  return tb - ta;
}

File: mobile/lib/inbox_tab_order.dart
import 'dart:convert';
import 'dart:ui' show Color;

import 'package:shared_preferences/shared_preferences.dart';

/// Chips de bandeja: Contactos · Grupos · No leídos · Favoritos (orden por usuario).
const kInboxTabIds = ['contacts', 'groups', 'unread', 'favorites'];

const kInboxTabLabels = {
  'contacts': 'Contactos',
  'groups': 'Grupos',
  'unread': 'No leídos',
  'favorites': 'Favoritos',
};

const _legacyMap = {
  'all': 'contacts',
  'contacts': 'contacts',
  'groups': 'groups',
  'unread': 'unread',
  'favorites': 'favorites',
};

String inboxTabOrderStorageKey(String? userId) =>
    'tacticalptx_chat_tab_order_${userId ?? 'anon'}';

List<String> normalizeInboxTabOrder(Iterable<dynamic>? raw) {
  final seen = <String>{};
  final out = <String>[];
  for (final id in raw ?? const []) {
    final mapped = _legacyMap[id?.toString()];
    if (mapped == null || seen.contains(mapped)) continue;
    seen.add(mapped);
    out.add(mapped);
  }
  for (final id in kInboxTabIds) {
    if (!seen.contains(id)) out.add(id);
  }
  return out;
}

Future<List<String>> loadInboxTabOrder(String? userId) async {
  try {
    final prefs = await SharedPreferences.getInstance();
    final raw = prefs.getString(inboxTabOrderStorageKey(userId));
    if (raw == null || raw.isEmpty) return List<String>.from(kInboxTabIds);
    final decoded = jsonDecode(raw);
    return normalizeInboxTabOrder(decoded is List ? decoded : null);
  } catch (_) {
    return List<String>.from(kInboxTabIds);
  }
}

Future<void> saveInboxTabOrder(String? userId, List<String> order) async {
  try {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(
      inboxTabOrderStorageKey(userId),
      jsonEncode(normalizeInboxTabOrder(order)),
    );
  } catch (_) {}
}

List<String> moveInboxTab(List<String> order, String fromId, String toId) {
  final next = normalizeInboxTabOrder(order);
  final from = next.indexOf(fromId);
  final to = next.indexOf(toId);
  if (from < 0 || to < 0 || from == to) return next;
  next.removeAt(from);
  next.insert(to, fromId);
  return next;
}

int compareContactRows({
  required bool aOnline,
  required bool bOnline,
  required int aGrade,
  required int bGrade,
  String? aAt,
  String? bAt,
  String? aPresence,
  String? bPresence,
}) {
  final ar = presenceSortRank(aPresence, aOnline);
  final br = presenceSortRank(bPresence, bOnline);
  if (ar != br) return ar.compareTo(br);
  if (aGrade != bGrade) return aGrade.compareTo(bGrade);
  final ta = aAt != null ? DateTime.tryParse(aAt)?.millisecondsSinceEpoch ?? 0 : 0;
  final tb = bAt != null ? DateTime.tryParse(bAt)?.millisecondsSinceEpoch ?? 0 : 0;
  return tb.compareTo(ta);
}

/// 0 en línea → 1 ausente → 2 desconectado → 3 fuera de línea.
int presenceSortRank(String? presence, bool online) {
  switch (normalizePresenceKey(presence, online)) {
    case 'online':
      return 0;
    case 'away':
      return 1;
    case 'stale':
      return 2;
    case 'offline':
    default:
      return 3;
  }
}

String normalizePresenceKey(String? presence, bool online) {
  final p = (presence ?? '').toLowerCase().trim();
  if (p == 'away' || p == 'background' || p == 'service') return 'away';
  if (p == 'online' ||
      p == 'radio' ||
      p == 'active' ||
      p == 'foreground') {
    return 'online';
  }
  if (p == 'stale') return 'stale';
  if (p == 'offline') return 'offline';
  return online ? 'online' : 'offline';
}

/// Colores semáforo (paridad GPS / mapa).
Color presenceSemaphoreColor(String? presence, {bool online = false}) {
  switch (normalizePresenceKey(presence, online)) {
    case 'online':
      return const Color(0xFF1F5A2E);
    case 'away':
      return const Color(0xFFA16207);
    case 'stale':
      return const Color(0xFF991B1B);
    case 'offline':
    default:
      return const Color(0xFF6B7280);
  }
}

String presenceSemaphoreLabel(String? presence, {bool online = false}) {
  switch (normalizePresenceKey(presence, online)) {
    case 'online':
      return 'En línea';
    case 'away':
      return 'Ausente';
    case 'stale':
      return 'Desconectado';
    case 'offline':
    default:
      return 'Fuera de línea';
  }
}


int compareByLastMessage(String? aAt, String? bAt) {
  final ta = aAt != null ? DateTime.tryParse(aAt)?.millisecondsSinceEpoch ?? 0 : 0;
  final tb = bAt != null ? DateTime.tryParse(bAt)?.millisecondsSinceEpoch ?? 0 : 0;
  return tb.compareTo(ta);
}
